File: rust/prophet-wasm/src/additional_features.rs
use crate::fourier::{FourierError, MAX_DENSE_ELEMENTS, checked_element_count};

/// Borrowed checked row-major additional-feature values.
#[derive(Clone, Copy, Debug)]
pub struct FeatureMatrixView<'a> {
  /// Number of rows.
  pub row_count: usize,
  /// Number of feature columns.
  pub column_count: usize,
  /// Row-major finite values.
  pub values: &'a [f64],
}

/// Borrowed priors and contiguous component ranges for additional columns.
#[derive(Clone, Copy, Debug)]
pub struct AdditionalFeatureLayoutView<'a> {
  /// One positive finite prior scale per feature column.
  pub prior_scales: &'a [f64],
  /// One starting column per named component.
  pub component_offsets: &'a [usize],
  /// One positive column count per named component.
  pub component_counts: &'a [usize],
}

/// Borrowed row-major binary mask with one column per seasonal component.
#[derive(Clone, Copy, Debug)]
pub struct SeasonalityMaskView<'a> {
  /// Number of timestamp rows.
  pub row_count: usize,
  /// Number of ordered seasonal components.
  pub component_count: usize,
  /// Row-major values, each exactly zero or one.
  pub values: &'a [u8],
}

/// Validated additional-feature and mask failures.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum AdditionalFeatureError {
  /// Matrix dimensions or aligned lengths differ.
  InvalidDimensions,
  /// A matrix value or prior scale is invalid.
  InvalidValue,
  /// Component ranges are not contiguous exact coverage.
  InvalidLayout,
  /// Dense dimensions overflow or exceed the memory policy.
  SizeOverflow,
}

impl FeatureMatrixView<'_> {
  /// Validate dimensions and finite matrix values.
  pub fn validate(&self, expected_rows: usize) -> Result<(), AdditionalFeatureError> {
    let expected = self
      .row_count
      .checked_mul(self.column_count)
      .ok_or(AdditionalFeatureError::SizeOverflow)?;

    if self.row_count != expected_rows || self.values.len() != expected {
      return Err(AdditionalFeatureError::InvalidDimensions);
    }

    if expected > MAX_DENSE_ELEMENTS {
      return Err(AdditionalFeatureError::SizeOverflow);
    }

    if self.values.iter().any(|value| !value.is_finite()) {
      return Err(AdditionalFeatureError::InvalidValue);
    }

    Ok(())
  }
}

impl AdditionalFeatureLayoutView<'_> {
  /// Validate priors and exact contiguous component coverage.
  pub fn validate(&self, column_count: usize) -> Result<(), AdditionalFeatureError> {
    if self.prior_scales.len() != column_count
      || self.component_offsets.len() != self.component_counts.len()
      || self
        .prior_scales
        .iter()
        .any(|value| !value.is_finite() || *value <= 0.0)
    {
      return Err(AdditionalFeatureError::InvalidValue);
    }

    let mut expected_offset = 0_usize;

    for (&offset, &count) in self.component_offsets.iter().zip(self.component_counts) {
      if offset != expected_offset || count == 0 {
        return Err(AdditionalFeatureError::InvalidLayout);
      }

      expected_offset = expected_offset
        .checked_add(count)
        .ok_or(AdditionalFeatureError::SizeOverflow)?;
    }

    if expected_offset != column_count {
      return Err(AdditionalFeatureError::InvalidLayout);
    }

    Ok(())
  }
}

impl SeasonalityMaskView<'_> {
  /// Validate mask dimensions and binary values.
  pub fn validate(
    &self,
    expected_rows: usize,
    expected_components: usize,
  ) -> Result<(), AdditionalFeatureError> {
    let expected = self
      .row_count
      .checked_mul(self.component_count)
      .ok_or(AdditionalFeatureError::SizeOverflow)?;

    if self.row_count != expected_rows
      || self.component_count != expected_components
      || self.values.len() != expected
    {
      return Err(AdditionalFeatureError::InvalidDimensions);
    }

    if expected > MAX_DENSE_ELEMENTS {
      return Err(AdditionalFeatureError::SizeOverflow);
    }

    if self.values.iter().any(|value| *value != 0 && *value != 1) {
      return Err(AdditionalFeatureError::InvalidValue);
    }

    Ok(())
  }
}

/// Return an owned all-one mask for unconditional components.
pub fn unconditional_masks(
  row_count: usize,
  component_count: usize,
) -> Result<Vec<u8>, AdditionalFeatureError> {
  let count = checked_element_count(row_count, component_count).map_err(|error| match error {
    FourierError::SizeOverflow | FourierError::SizeLimitExceeded => {
      AdditionalFeatureError::SizeOverflow
    }
    _ => AdditionalFeatureError::InvalidDimensions,
  })?;

  Ok(vec![1; count])
}

/// Evaluate grouped additional-feature contributions in row-major order.
pub fn evaluate_additional_components(
  matrix: FeatureMatrixView<'_>,
  layout: AdditionalFeatureLayoutView<'_>,
  coefficients: &[f64],
) -> Result<Vec<f64>, AdditionalFeatureError> {
  matrix.validate(matrix.row_count)?;
  layout.validate(matrix.column_count)?;

  if coefficients.len() != matrix.column_count
    || coefficients.iter().any(|value| !value.is_finite())
  {
    return Err(AdditionalFeatureError::InvalidDimensions);
  }

  let output_count = checked_element_count(matrix.row_count, layout.component_offsets.len())
    .map_err(|_| AdditionalFeatureError::SizeOverflow)?;
  let mut output = vec![0.0; output_count];

  for row in 0..matrix.row_count {
    for component in 0..layout.component_offsets.len() {
      let offset = layout.component_offsets[component];
      let count = layout.component_counts[component];
      let mut value = 0.0;

      for (column, &coefficient) in coefficients.iter().enumerate().skip(offset).take(count) {
        value += matrix.values[row * matrix.column_count + column] * coefficient;
      }

      if !value.is_finite() {
        return Err(AdditionalFeatureError::InvalidValue);
      }

      output[row * layout.component_offsets.len() + component] = value;
    }
  }

  Ok(output)
}

File: src/internal/additional-features.ts
import { Effect, Schema } from "effect";

import type { FeatureName } from "../feature-name";

/** Source kind represented by one contiguous additional-feature component. */
export type AdditionalFeatureKind = "event" | "regressor";

/** One named contiguous range in the additional coefficient block. */
export interface AdditionalFeatureComponent {
  readonly kind: AdditionalFeatureKind;
  readonly name: FeatureName;
  readonly coefficientOffset: number;
  readonly coefficientCount: number;
}

/** Ordered metadata and priors for additional additive columns. */
export interface AdditionalFeatureLayout {
  readonly components: ReadonlyArray<AdditionalFeatureComponent>;
  readonly coefficientCount: number;
  readonly priorScales: ReadonlyArray<number>;
}

/** A privately owned, checked row-major matrix of finite additional features. */
export interface AdditionalFeatureMatrix {
  readonly rowCount: number;
  readonly columnCount: number;
  readonly values: Float64Array;
}

/** A privately owned row-major binary mask with one column per seasonal component. */
export interface SeasonalityMaskMatrix {
  readonly rowCount: number;
  readonly componentCount: number;
  readonly values: Uint8Array;
}

/** Matrix and layout passed together to numerical backends. */
export interface KnownAdditiveFeatures {
  readonly matrix: AdditionalFeatureMatrix;
  readonly layout: AdditionalFeatureLayout;
}

/** Structured failure while constructing checked additional-feature values. */
export class InvalidAdditionalFeatures extends Schema.TaggedError<InvalidAdditionalFeatures>()(
  "InvalidAdditionalFeatures",
  {
    path: Schema.Array(Schema.PropertyKey),
    message: Schema.String,
  },
) {}

const fail = (path: ReadonlyArray<PropertyKey>, message: string) =>
  Effect.fail(new InvalidAdditionalFeatures({ path, message }));

const checkedElementCount = (rows: number, columns: number): number | undefined => {
  if (!Number.isSafeInteger(rows) || rows < 0 || !Number.isSafeInteger(columns) || columns < 0) {
    return undefined;
  }

  const count = rows * columns;

  return Number.isSafeInteger(count) ? count : undefined;
};

/** Construct and deeply freeze an ordered additional-feature layout. */
export const createAdditionalFeatureLayout = <Component extends AdditionalFeatureComponent>(
  components: ReadonlyArray<Component>,
  priorScales: ReadonlyArray<number>,
): Effect.Effect<
  AdditionalFeatureLayout & { readonly components: ReadonlyArray<Component> },
  InvalidAdditionalFeatures
> => {
  const names = new Set<string>();
  let expectedOffset = 0;

  for (const [index, component] of components.entries()) {
    if (names.has(component.name)) {
      return fail(["components", index, "name"], `Feature name '${component.name}' is duplicated`);
    }

    names.add(component.name);

    if (component.coefficientOffset !== expectedOffset) {
      return fail(
        ["components", index, "coefficientOffset"],
        `Expected contiguous coefficient offset ${expectedOffset}`,
      );
    }

    if (!Number.isSafeInteger(component.coefficientCount) || component.coefficientCount <= 0) {
      return fail(
        ["components", index, "coefficientCount"],
        "Component coefficient count must be a positive safe integer",
      );
    }

    if (expectedOffset > Number.MAX_SAFE_INTEGER - component.coefficientCount) {
      return fail(["components", index], "Additional coefficient count exceeds safe arithmetic");
    }

    expectedOffset += component.coefficientCount;
  }

  if (priorScales.length !== expectedOffset) {
    return fail(["priorScales"], `Expected exactly ${expectedOffset} additional prior scales`);
  }

  for (const [index, priorScale] of priorScales.entries()) {
    if (!Number.isFinite(priorScale) || priorScale <= 0) {
      return fail(["priorScales", index], "Additional prior scales must be positive and finite");
    }
  }

  const frozenComponents: ReadonlyArray<Component> = Object.freeze(
    components.map((component) => {
      // SAFETY: the copy preserves every field of Component while severing the caller-owned object alias.
      const copied = { ...component } as Component;

      return Object.freeze(copied);
    }),
  );

  return Effect.succeed(
    Object.freeze({
      components: frozenComponents,
      coefficientCount: expectedOffset,
      priorScales: Object.freeze(Array.from(priorScales)),
    }),
  );
};

/** Construct an owned checked row-major additional-feature matrix. */
export const createAdditionalFeatureMatrix = (
  rowCount: number,
  columnCount: number,
  values: ArrayLike<number>,
): Effect.Effect<AdditionalFeatureMatrix, InvalidAdditionalFeatures> => {
  const expectedLength = checkedElementCount(rowCount, columnCount);

  if (expectedLength === undefined) {
    return fail([], "Additional feature matrix dimensions exceed safe arithmetic");
  }

  if (values.length !== expectedLength) {
    return fail(["values"], `Expected exactly ${expectedLength} matrix values`);
  }

  const owned = new Float64Array(values);

  for (const [index, value] of owned.entries()) {
    if (!Number.isFinite(value)) {
      return fail(["values", index], "Additional feature values must be finite");
    }
  }

  return Effect.succeed(Object.freeze({ rowCount, columnCount, values: owned }));
};

/** Construct an owned checked row-major seasonal-component mask. */
export const createSeasonalityMaskMatrix = (
  rowCount: number,
  componentCount: number,
  values: ArrayLike<number>,
): Effect.Effect<SeasonalityMaskMatrix, InvalidAdditionalFeatures> => {
  const expectedLength = checkedElementCount(rowCount, componentCount);

  if (expectedLength === undefined) {
    return fail([], "Seasonality mask dimensions exceed safe arithmetic");
  }

  if (values.length !== expectedLength) {
    return fail(["values"], `Expected exactly ${expectedLength} mask values`);
  }

  const owned = new Uint8Array(expectedLength);

  for (let index = 0; index < expectedLength; index += 1) {
    const value = values[index];

    if (value !== 0 && value !== 1) {
      return fail(["values", index], "Seasonality masks must contain exactly zero or one");
    }

    owned[index] = value;
  }

  return Effect.succeed(Object.freeze({ rowCount, componentCount, values: owned }));
};

/** Construct an all-one mask for unconditional seasonal components. */
export const createUnconditionalSeasonalityMask = (
  rowCount: number,
  componentCount: number,
): Effect.Effect<SeasonalityMaskMatrix, InvalidAdditionalFeatures> => {
  const count = checkedElementCount(rowCount, componentCount);

  if (count === undefined) {
    return fail([], "Seasonality mask dimensions exceed safe arithmetic");
  }

  const values = new Uint8Array(count);
  values.fill(1);

  return createSeasonalityMaskMatrix(rowCount, componentCount, values);
};

/** Concatenate checked additional-feature blocks while preserving row and component order. */
export const concatenateKnownAdditiveFeatures = (
  blocks: ReadonlyArray<KnownAdditiveFeatures>,
): Effect.Effect<KnownAdditiveFeatures, InvalidAdditionalFeatures> =>
  Effect.gen(function* () {
    const rowCount = blocks[0]?.matrix.rowCount ?? 0;
    let columnCount = 0;

    for (const [index, block] of blocks.entries()) {
      if (block.matrix.rowCount !== rowCount) {
        return yield* fail(
          [index, "matrix", "rowCount"],
          "Additional feature blocks must have identical row counts",
        );
      }

      if (columnCount > Number.MAX_SAFE_INTEGER - block.matrix.columnCount) {
        return yield* fail([], "Combined additional feature dimensions exceed safe arithmetic");
      }

      columnCount += block.matrix.columnCount;
    }

    const elementCount = checkedElementCount(rowCount, columnCount);

    if (elementCount === undefined) {
      return yield* fail([], "Combined additional feature dimensions exceed safe arithmetic");
    }

    const values = new Float64Array(elementCount);
    const components: Array<AdditionalFeatureComponent> = [];
    const priorScales: Array<number> = [];
    let columnOffset = 0;

    for (const block of blocks) {
      for (let row = 0; row < rowCount; row += 1) {
        const sourceOffset = row * block.matrix.columnCount;
        const targetOffset = row * columnCount + columnOffset;

        values.set(
          block.matrix.values.subarray(sourceOffset, sourceOffset + block.matrix.columnCount),
          targetOffset,
        );
      }

      for (const component of block.layout.components) {
        components.push({
          ...component,
          coefficientOffset: component.coefficientOffset + columnOffset,
        });
      }

      priorScales.push(...block.layout.priorScales);
      columnOffset += block.matrix.columnCount;
    }

    const layout = yield* createAdditionalFeatureLayout(components, priorScales);
    const matrix = yield* createAdditionalFeatureMatrix(rowCount, columnCount, values);

    return Object.freeze({ matrix, layout });
  });

/** Construct the canonical empty additional-feature values for a row count. */
export const createEmptyKnownAdditiveFeatures = (
  rowCount: number,
): Effect.Effect<KnownAdditiveFeatures, InvalidAdditionalFeatures> =>
  Effect.gen(function* () {
    const layout = yield* createAdditionalFeatureLayout([], []);
    const matrix = yield* createAdditionalFeatureMatrix(rowCount, 0, []);

    return Object.freeze({ matrix, layout });
  });
